tools/rostest/nodes/publishtest

```diff
@@ -107,10 +108,15 @@ class PublishTest(unittest.TestCase):
     def test_publish(self):
         """Test topics are published and messages come"""
         use_sim_time = rospy.get_param('/use_sim_time', False)
-        while use_sim_time and (rospy.Time.now() == rospy.Time(0)):
+        t_start = time.time()
+        while not rospy.is_shutdown() and \
+                use_sim_time and (rospy.Time.now() == rospy.Time(0)):
             rospy.logwarn('/use_sim_time is specified and rostime is 0, '
                           '/clock is published?')
-            rospy.sleep(0.1)
+            if time.time() - t_start > 10:
+                self.fail('Timed out (10s) of /clock publication.')
+            # must use time.sleep because /clock isn't yet published, so rospy.sleep hangs.
+            time.sleep(1)
```

Comment:
In that case, the warning should only be printed once to avoid flooding the console unnecessarily.